### ISSUE TITLE
definition for PRERELEASE and BUILD doesn't conform to semver 2.0.0

### ISSUE THREAD

**Issue by qyanu** [fsaintjacques/semver-tool#29]
Hi François,

in https://github.com/fsaintjacques/semver-tool/blob/eb430f81c353feee4c4ab8206f0844d283e498dd/src/semver
in line 23 and following, the definition for PRERELEASE is given as follows:
> PRERELEASE is an optional string composed of alphanumeric characters and hyphens

and the definition for BUILD is given as follows:
> BUILD is also an optional string composed of alphanumeric characters and hyphens

The regex in line 5 of the same file seems to match these definitions.

However, the README.md in the repositorie's root folder explicitely declares compatibility with semver as defined by https://github.com/mojombo/semver which gives a slighty different definition for PRERELEASE and BUILD respectively. Quoted from there:
> a series of dot separated identifiers
and further details the notion of "identifier" as follows:
> Identifiers MUST comprise only ASCII alphanumerics and hyphen [0-9A-Za-z-].
> Identifiers MUST NOT be empty.
> Numeric identifiers MUST NOT include leading zeroes.

The definition as given by semver is at odds with the definition given by this script. I would consider this a bug (and therefore this bug report) and propose the following options of resolving it:

a) documentation and code-comments should state clearly that the difference is intended. preferably a reason or reasons should also be given as to why the difference exists.
b) the README.md in the root of this repository should explicitely reference semver 1.x, as this version has a different definition of PRERELEASE that is more similar to this script's implementation.
c) change this script's implementation in order to match the semver 2.x definition.

With regards to (a), i could not be of any help as i do not have knowledge of the development process that lead up to the current state.
Option (b) seems to be the quickest solution with regards to implementation effort. However, semver 2.x was released quite some time ago and is the de-facto norm. It seems to me, that when people make unqualified reference to "semver", that version 2 is meant.

Regarding option (c), i intend to attach a patch to this issue shortly.

with best wishes,

Max

**Comment by fsaintjacques** [comment]
go ahead with c), it would be nice if you could provide a matrix of what is/was accepted/rejected.

**Comment by qyanu** [comment]
I just wanted to edit the source, as i discovered that the regex already accepts dots as parts of PRERELEASE and BUILD. Maybe the patch will reduce itself to just change the wording of the documentation.

**Comment by qyanu** [comment]
I think the patch attached is to the best of my abilities (sadly, i must say).

**Comment by qyanu** [comment]
To summarize: Indeed i couldn't find any combination that didn't pass unit testing.

I wasn't convinced that my idea to handle error-cases in unit-tests is the best, so i didn't include it.
My idea would be this: reserve some special return value that's impossible normally. For example the value "@error". Write this value into the README.md in place where the expected result would be written. Then the script test/documentation-test could be extended detect this value and then epect a non-zero exit status from semver.

Regarding the specified comparison rules of semver2.x with regards to prerel, i am not confident enough to say there's a bug without testing. but i didn't want to test without having the tests documented and repeatable -- thus the above mentioned feature of test/documentation-test would need to be resolved first.

However, i must be frank, that i could not guarantee time for such endavours. therefore i will only promise to go through with the changes as proposed (and discuss any problems you might find).

**Comment by ranger6** [comment]
As I had already opened the code box to fix issue #33, I decided to jump in and work on semver spec 2.0.0.
There is already an open pull request (#30) addressing this issue, so I don't want to yank this contribution around.  Herr Pogner ( @qyanu), if you have moved forward on the pull request or plan to shortly, let us know.
On the other hand, if your PR is still at the 1-Aug-2019 state and is not going to more forward (you hinted above about time constraints and test-case doubts), I'm ready to publish a PR that should close this bug.
1. It complies with 2.0.0.
2. Automated unit tests show previous non-compliance as well as testing compliance. Basis for the "matrix of what is/was accepted/rejected" requested by @fsaintjacques.
3. No changes to test/documentation-test.
4. No CLI changes.
I'll be publishing my work shortly and will send a PR unless you'd rather I don't.
